flw/src/main/java/org/motechproject/nms/flw/service/impl/FrontLineWorkerServiceImpl.java

```diff
@@ -150,6 +163,69 @@ public State getState(FrontLineWorker frontLineWorker) {
         return state;
     }
 
+    @Override // NO CHECKSTYLE Cyclomatic Complexity
+    public boolean createUpdate(Map<String, Object> flw) {
+
+        long stateId = (long) flw.get(FlwConstants.STATE_ID);
+        long districtId = (long) flw.get(FlwConstants.DISTRICT_ID);
+        String mctsFlwId = flw.get(FlwConstants.ID).toString();
+        long contactNumber = (long) flw.get(FlwConstants.CONTACT_NO);
+
+        State state = locationService.getState(stateId);
+        if (state == null) {
+            flwErrorDataService.create(new FlwError(mctsFlwId, stateId, districtId, FlwErrorReason.INVALID_LOCATION_STATE));
+            return false;
+        }
+        District district = locationService.getDistrict(stateId, districtId);
+        if (district == null) {
+            flwErrorDataService.create(new FlwError(mctsFlwId, stateId, districtId, FlwErrorReason.INVALID_LOCATION_DISTRICT));
+            return false;
+        }
+
+        FrontLineWorker existingFlwByNumber = getByContactNumber(contactNumber);
+        FrontLineWorker existingFlwByMctsFlwId = getByMctsFlwIdAndState(mctsFlwId, state);
+        Map<String, Object> location = new HashMap<>();
+        try {
+            location = locationService.getLocations(flw);
+
+            if (existingFlwByMctsFlwId != null && existingFlwByNumber != null) {
+
+                if (existingFlwByMctsFlwId.getMctsFlwId().equalsIgnoreCase(existingFlwByNumber.getMctsFlwId()) &&
+                        existingFlwByMctsFlwId.getState().equals(existingFlwByNumber.getState())) {
+                    // we are trying to update the same existing flw. set fields and update
+                    LOGGER.debug("Updating existing user with same phone number");
+                    update(FlwMapper.updateFlw(existingFlwByMctsFlwId, flw, location));
+                    return true;
+                } else {
+                    // we are trying to update 2 different users and/or phone number used by someone else
+                    LOGGER.debug("Existing flw but phone number(update) already in use");
+                    flwErrorDataService.create(new FlwError(mctsFlwId, stateId, districtId, FlwErrorReason.PHONE_NUMBER_IN_USE));
+                    return false;
+                }
+            } else if (existingFlwByMctsFlwId != null && existingFlwByNumber == null) {
+                // trying to update the phone number of the person. possible migration scenario
+                // making design decision that flw will lose all progress when phone number is changed. Usage and tracking is not
+                // worth the effort & we don't really know that its the same flw
+                LOGGER.debug("Updating phone number for flw");
+                update(FlwMapper.updateFlw(existingFlwByMctsFlwId, flw, location));
+                return true;
+            } else if (existingFlwByMctsFlwId == null && existingFlwByNumber != null) {
+                // phone number used by someone else.
```

Comment:
Is it guarenteed to be the case that the phone number is in use by someone else?  Could we have just had the FLW without their mcts flw id?